org-ai-openai-image-query.el

```diff
@@ -0,0 +1,124 @@
+  "The model to use for image queries."
+  :group 'org-ai
+  :type '(choice (const :tag "gpt-4o-mini" "gpt-4o-mini")
+          (const :tag "gpt-4o" "gpt-4o")
+          (const :tag "gpt-4-turbo" "gpt-4-turbo")))
+
+(defcustom org-ai-query-image-file nil
+  "Optional file used to store the `org-ai-query-image' conversations in.
+If nil, a buffer with no file backing is used. If a file is
+specified, new conversations are appended to the file or
+function's output. Function should take no arguments and return a
+filename."
+  :group 'org-ai
+  :type '(choice (const :tag "No file" nil)
+          (file :tag "File")
+          (function :tag "Function")))
+
+(defvar org-ai-openai-image-query-endpoint "https://api.openai.com/v1/chat/completions"
+  "Endpoint for querying images with OpenAI.")
+
+(defun org-ai--send-image-query (image-content question callback)
+  "Send IMAGE-CONTENT and QUESTION to the OpenAI API.
+Calls CALLBACK with the response."
+  (let ((url-request-extra-headers `(("Authorization" . ,(string-join `("Bearer" ,(org-ai--openai-get-token)) " "))
+                                     ("Content-Type" . "application/json")))
+        (url-request-method "POST")
+        (url-request-data (json-encode `(("model" . ,org-ai-image-query-model)
+                                         ("messages" . [(("role" . "user")
+                                                         ("content" . [((type . "text") (text . ,question))
+                                                                       ,image-content]))])
+                                         ("max_tokens" . ,(or org-ai-default-max-tokens 300))))))
+    (url-retrieve
+     org-ai-openai-image-query-endpoint
+     (lambda (_status)
+       (goto-char url-http-end-of-headers)
+       (let ((response (json-read)))
+         (if response
+             (funcall callback response)
+           (message "Failed to get a valid response from OpenAI API.")))))))
+
+(defun org-ai--send-base64-image-query (base64-image question callback)
+  "Send a BASE64-IMAGE and QUESTION to the OpenAI API.
+Calls CALLBACK with the response."
+  (org-ai--send-image-query
+   `((type . "image_url")
+     (image_url . (("url" . ,(concat "data:image/jpeg;base64," base64-image)))))
+   question callback))
+
+(defun org-ai--send-url-image-query (image-url question callback)
+  "Send an IMAGE-URL and QUESTION to the OpenAI API.
+Calls CALLBACK with the response."
+  (org-ai--send-image-query
+   `((type . "image_url")
+     (image_url . (("url" . ,image-url))))
+   question callback))
+
+(defun org-ai--handle-openai-response (response)
+  "Handle the RESPONSE from OpenAI API."
+  (let* ((choices (alist-get 'choices response))
+         ;; Ensure choices is a vector and has at least one element
+         (choice (and (vectorp choices)
+                      (> (length choices) 0)
+                      (aref choices 0)))
+         (message_ (alist-get 'message choice))
```

Comment:
this all feels brittle, should consider extracting this into a separate function so when/if it breaks it's even more isolated and fixing it is easier